Feat move to waiting children for manual processing

roggervalf opened the pull request:
Trying to get advantage of the new state of waiting children, so when creating a job we can relate a parent, then we can decide in a parent flow when to wait for children with this new method.
Description:
moveToWaitingChildren: allows you to move a job from active state to waiting-children. In manual processing, you can decide when to wait for those children to be completed as you need. Remember that you cannot finished a job if there are children jobs that are not completed.
Note: in order to create children jobs in a manual way, you can pass the parent option when creating one, the parent job could still be processed and decide when to wait for those children with moveToWaitingChildren.

manast commented on src/classes/job.ts:
```diff
@@ -528,14 +533,22 @@ export class Job<T = any, R = any, N extends string = string> {
     const queue = this.queue;
 
     const jobData = this.asJSON();
+    const parentDependenciesKey = this.opts.parent
```
we can now use ```this.parentKey``` which is already the same as `${this.opts.parent.queue}:${this.opts.parent.id` we could even deprecate the parent opt soon since it is now superfluous.

roggervalf replied:
hey, I'm seeing that for .create the only way for now to obtain the parentKey is with opts, no sure if this should be refactored here, since await Job.create does not have this.parentKey when creating a job. What do you think about it?

roggervalf replied:
In last version I reused getParentKey to refactor this, now parentKey could be assigned in job constructor

manast reviewed (changes requested):
Thanks for the PR.
 I gave you some comments.
It would be great if you could provide a description of the complete usecase since I will need to document how this feature is used and I am not grasping all the details.
 I get that this method can be used within the process function of a worker, so that you can go from being active to waiting a series of children. But I have 2 questions:
1) how to do avoid that the job that is being processed and moved to "wait-children" is not moved to the "completed" set cleanly? I suspect you get an error now since the job is not in the active status when the processor ends and tries to "complete" the job.
2) When and how do you define the children? I think that moveToWaitingChildren should be an implicit operation when actually adding the children, otherwise there is a risk that you manage to move the job to waiting-children but fail to actually add the children and so you will end with a dangling parent job that can never be processed.

roggervalf commented:
> Thanks for the PR.
> I gave you some comments.
> It would be great if you could provide a description of the complete usecase since I will need to document how this feature is used and I am not grasping all the details.
> I get that this method can be used within the process function of a worker, so that you can go from being active to waiting a series of children. But I have 2 questions:
> 
>     1. how to do avoid that the job that is being processed and moved to "wait-children" is not moved to the "completed" set cleanly? I suspect you get an error now since the job is not in the active status when the processor ends and tries to "complete" the job.
> 
>     2. When and how do you define the children? I think that moveToWaitingChildren should be an implicit operation when actually adding the children, otherwise there is a risk that you manage to move the job to waiting-children but fail to actually add the children and so you will end with a dangling parent job that can never be processed.



> Thanks for the PR.
> I gave you some comments.
> It would be great if you could provide a description of the complete usecase since I will need to document how this feature is used and I am not grasping all the details.
> I get that this method can be used within the process function of a worker, so that you can go from being active to waiting a series of children. But I have 2 questions:
> 
>     1. how to do avoid that the job that is being processed and moved to "wait-children" is not moved to the "completed" set cleanly? I suspect you get an error now since the job is not in the active status when the processor ends and tries to "complete" the job.
> 
>     2. When and how do you define the children? I think that moveToWaitingChildren should be an implicit operation when actually adding the children, otherwise there is a risk that you manage to move the job to waiting-children but fail to actually add the children and so you will end with a dangling parent job that can never be processed.

sure, for your questions:
1. Yes, I'm moving the job to wait-children if I find dependencies, on the other hand it is not moved to wait-children and can continue with his job. When it is moved to waiting-children it is deleted from active.
2. In this case I'm checking if there are dependencies for the parent job, so as those are not processed the job can be moved, with this check we prevent to have dangling parent job, I add an example in a test where job are processed in a manual way, so children can be processed before the parent go to waiting-children.
From my use case I'm currently working with a parent job that can trigger in some steps children jobs. i.e:
Steps:
1. Step 1 do something
2. Step 2 create a child job (could more or less time to be processed than step 3)
3. Step 3 do another thing (don't depend on children jobs)
4. Step 4 create another child job (could more or less time to be processed than step 5)
5. Step 5 do another thing (don't depend on children jobs)
6. Step 6 waiting for children
7. Step 7 we already have children finished so we can use their results (depends on children results)
8. Step 8 completed
So with with moveToWaitingChildren we can have the flexibility to wait for children and maybe have step 3 and 5 already done, sometimes we depend on jobs can take a lot of time to be completed so would be nice to save for the parent to also processed steps that don't depend on those children.

manast commented on src/commands/moveToWaitingChildren-4.lua:
```diff
@@ -0,0 +1,42 @@
+end
+
+if rcall("EXISTS", KEYS[4]) == 1 then
+  if rcall("SCARD", KEYS[4] .. ":dependencies") ~= 0 then 
```
I do not grasp why this check is needed?. As I understand it, it only moves the job if there are some dependencies already for that job, but shouldn't we check instead if this particular dependency is part of the dependency list?

roggervalf replied:
oh, here it is to move the parent, I'm checking if there is any dependency in the dependency list of this parent job

roggervalf replied:
also this gives another feature that I can add, as you mentioned we can also have a method to check if a particular dependency is part of the dependencies list. For now this method is what my team currently is kind of emulating, waiting for all the children at some point. Let me think how to refactor this to also support on waiting for a particular dependency that is passed on

manast commented on src/test/test_job.ts:
```diff
@@ -61,6 +61,84 @@ describe('Job', function() {
+          const { processed } = await job.getDependencies();
+
+          if (Object.keys(processed).length !== 3) {
+            expect(movedToWaitingChildren).to.be.true;
```
hmm, in this case the promise will never resolve.

roggervalf replied:
let me refactor this test

manast commented on src/test/test_job.ts:
```diff
@@ -61,6 +61,84 @@ describe('Job', function() {
+    const processingParent = new Promise<void>((resolve, reject) => [
+      (parentProcessor = async (job: Job) => {
+        try {
+          const movedToWaitingChildren = await job.moveToWaitingChildren('0');
```
why is '0' the token and not parentWorker.token ?

roggervalf replied:
tyty I was not sure where to get a token in this case

manast replied:
ok. However in this case we really need the token from the worker that picked the job, I am surprised this is working since moveToWaitingChildren must match the stored token in the lock and the token passed to it. For this to work we need to pass the token to the processor, maybe as an extra parameter to the processor function.

manast commented on src/test/test_worker.ts:
```diff
@@ -1684,6 +1684,116 @@ describe('workers', function() {
+        unprocessed: unprocessed2,
+      } = await parent.getDependencies();
+      const movedToWaitingChildren = await parent.moveToWaitingChildren(
+        parentToken,
```
parentToken should be worker.token, and anything else different from worker.token should raise an exception.

roggervalf replied:
I'm reusing the parentToken that I passed to the worker when calling getNextJob, if I use parentWorker.token it gets an error of **Property 'token' does not exist on type 'Worker<any, any, string>'. Did you mean 'toKey'?**
![Selection_024](https://user-images.githubusercontent.com/14021523/115286113-50167900-a114-11eb-8901-2fce76d21899.png)

manast replied:
ah sorry, you are right, token is private and when getting jobs manually you must create the token yourself.

manast commented on src/test/test_job.ts:
```diff
@@ -61,6 +61,84 @@ describe('Job', function() {
+    const processingParent = new Promise<void>((resolve, reject) => [
+      (parentProcessor = async (job: Job) => {
+        try {
+          const movedToWaitingChildren = await job.moveToWaitingChildren('0');
```
I am still a bit confused that we can allow this call from inside the worker processor, when the call ends it will try to move the job to "completed" from "active", but since we manually moved the job to "waiting-children" it should raise an error somewhere.

roggervalf replied:
sure, in moveToFinished I'm also checking if there are pending dependencies, in case there are, throws an error

manast replied:
yes I saw that, but the current test code, does it not trigger that  error? since the job has been moved and when the processor call returns the worker will try to move the job to completed and your check should return an error. So two things here to think about, how to handle this error (I suspect it is erroring silently now), and two, what is the proper way to do this without raising an error. Somehow we have to tell the worker that we have moved the job to waiting-children so that it does not matter anymore if the job was in active, and in this case do a no op. Or maybe "moveToWaitingChildren" should be a method that can only be used when processing jobs manually. Do you have a need for this use case where it has to be called inside the processor function?

roggervalf commented:
@manast the pr was updated with master, please when you get some time :eyes:

manast commented on src/commands/moveToWaitingChildren-4.lua:
```diff
@@ -0,0 +1,60 @@
+  end
+end
+
+return -1
```
missing newline

manast commented:
Can you join for a conversation here: https://bullmq.slack.com/ I would like to discuss this PR in more detail.

roggervalf commented:
> Can you join for a conversation here: https://bullmq.slack.com/ I would like to discuss this PR in more detail.

sure, I guess I should be invited to that workspace :sweat_smile:. rogger.valverde@uni.pe

manast commented on src/test/test_job.ts:
```diff
@@ -63,6 +63,80 @@ describe('Job', function() {
+        const movedToWaitingChildren = await job.moveToWaitingChildren(token);
+        const { processed } = await job.getDependencies();
+
+        if (Object.keys(processed).length !== 3) {
```
so basically here, this will be true the first time this job is processed right? and then when it is processed the second time after all children are processed it will be false. Please correct me if my assumption was wrong.

roggervalf replied:
yes you are right, the first try it should have children, after it move waiting-children the second try will go to the else block

manast replied:
in both this cases there must be some silent errors since the first time the job fails but BullMQ is not able to move to job to failed, and in the other case it completes and then it works. I want to figure out an API where moving the processed job from within a processor is legit without breaking current semantics, since I see other uses cases that are important, for example being able for a job to delay itself if some condition is met during the processing.

roggervalf replied:
Sounds also like a good addition to support polling

roggervalf commented on src/classes/worker.ts:
```diff
@@ -326,12 +326,20 @@ export class Worker<
     lockExtender();
     try {
       const result = await this.processFn(job, token);
+      if (typeof result === 'string' && result === 'waiting-children') {
```
@manast, what do you think about this change?, we could send a reserved word to handle the moveToWaitingChildren in a worker processJob

manast replied:
I thought about this but I think it is a bit hackish. Either enhance the "result" object so that it is richer, like:
```ts
interface Result {
   data?: T,
   type: 'data' | 'waiting-children' | 'delay' | 'etc '
}
```
or having an internal job result type that is set up implicitly when calling job.moveToWaitingChildren, etc.
The problem with the first approach (that is probably the cleaner one), is that it will become a breaking change...

roggervalf replied:
About the second one, also I was thinking on setting a field in job object, I can try the second one

roggervalf commented:
@manast this pr is ready for another round xd please when you had a chance

manast commented:
This looks good to me for manually processing children. I can merge it if you do not have anything else to add. 
I wont be able to write about this in here: https://docs.bullmq.io/patterns/manually-fetching-jobs because this use case is still a bit foreign to me, so I will probably write something that is not reflecting the real use cases.

roggervalf commented:
> This looks good to me for manually processing children. I can merge it if you do not have anything else to add.
> I wont be able to write about this in here: https://docs.bullmq.io/patterns/manually-fetching-jobs because this use case is still a bit foreign to me, so I will probably write something that is not reflecting the real use cases.

Sure, it is ready
